### ISSUE TITLE
Walled Garden is not working for baidu.com

### ISSUE THREAD

**Issue by zedgu** [openNDS/openNDS#156]
Hardware: Raspberry Pi 4 Model B Rev 1.2
OS: OpenWrt R21.4.18
opennds: 8.1.1-1
ipset: 7.4-1
dnsmasq-full: 2.85-1

I'm using legacy html splash page, and `ndsctl status` like:
```shell
$ ndsctl status
==================
openNDS Status
====
Version: 8.1.1
Uptime: 2s
Gateway Name: CA
Managed interface: br-lan
Managed IP range: 0.0.0.0/0
Server listening: http://192.168.10.1:2050
Binauth: Disabled
Preauth: Disabled
FAS: Disabled
Client Check Interval: 60s
Rate Check Window: 2 check intervals (120s)
Preauthenticated Client Idle Timeout: 30m
Authenticated Client Idle Timeout: 120m
Download rate limit (default per client): no limit
Upload rate limit (default per client): no limit
Download quota (default per client): no limit
Upload quota (default per client): no limit
Total download: 1 kByte; avg: 6.83 kbit/s
Total upload: 9 kByte; avg: 39.42 kbit/s
====
Client authentications since start: 0
Current clients: 0
====
Blocked MAC addresses: none
Allowed MAC addresses: N/A
Trusted MAC addresses: none
Walled Garden FQDNs:
  baidu.com
Walled Garden Ports:
  80
  443
========
```

BUT , ping like:
```shell
ping baidu.com
PING baidu.com (39.156.69.79): 56 data bytes
Request timeout for icmp_seq 0
Request timeout for icmp_seq 1
Request timeout for icmp_seq 2
Request timeout for icmp_seq 3
```

am I missing any settings?

**Comment by bluewavenet** [comment]
@zedgu 
ndsctl status shows you are restricting Walled Garden access to ports 80 and 443.

Ping however, does not use a tcp/udp port, instead it uses the ICMP protocol, so will be blocked.

If you need to be able to ping baidu.com, then you must not set any walled garden ports, thus allowing access to baidu.com for all protocols.

Usually all that is required for preauthenticated clients in the Walled Garden is is web access ie ports 80 and 443.

**Comment by bluewavenet** [comment]
@zedgu 
I can check your configuration, so please show the output of:
`ipset list`

and

`uci show opennds`

**Comment by bluewavenet** [comment]
@zedgu 
I have investigated a little. It seems Baidu is structured to use multiple FQDNs as well as direct IP address access.

You will have to list not only all the FQDNs it uses, but also manually add the ip addresses it uses without DNS.
Usually the same FQDN and IP address information is used to block access, whereas you are trying to ALLOW access.

I suspect Baidu have done this intentionally to stop blocking - after all search engines generate a huge amount of revenue.
**If I am correct, the problem is, this will also stop you from allowing access in your walled garden.**

**Comment by zedgu** [comment]
> @zedgu
> I can check your configuration, so please show the output of:
> `ipset list`
> 
> and
> 
> `uci show opennds`

There's too mang IP ranges before, and I restarted opennds after cleanning all the IPs by `ipset flush`. But the problem is still the same.

```shell
$ ipset list                 
Name: cn
Type: hash:net
Revision: 6
Header: family inet hashsize 2048 maxelem 65536
Size in memory: 480
References: 0
Number of entries: 0
Members:

Name: ct
Type: hash:net
Revision: 6
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 456
References: 0
Number of entries: 0
Members:

Name: cnc
Type: hash:net
Revision: 6
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 456
References: 0
Number of entries: 0
Members:

Name: cmcc
Type: hash:net
Revision: 6
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 456
References: 0
Number of entries: 0
Members:

Name: crtc
Type: hash:net
Revision: 6
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 456
References: 0
Number of entries: 0
Members:

Name: cernet
Type: hash:net
Revision: 6
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 456
References: 0
Number of entries: 0
Members:

Name: gwbn
Type: hash:net
Revision: 6
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 456
References: 0
Number of entries: 0
Members:

Name: othernet
Type: hash:net
Revision: 6
Header: family inet hashsize 2048 maxelem 65536
Size in memory: 480
References: 0
Number of entries: 0
Members:

Name: walledgarden
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 208
References: 4
Number of entries: 0
Members:

Name: music
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 208
References: 0
Number of entries: 0
Members:

Name: mwan3_connected_v4
Type: hash:net
Revision: 6
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 456
References: 0
Number of entries: 0
Members:

Name: mwan3_connected_v6
Type: hash:net
Revision: 6
Header: family inet6 hashsize 1024 maxelem 65536
Size in memory: 1240
References: 0
Number of entries: 0
Members:

Name: mwan3_source_v6
Type: hash:net
Revision: 6
Header: family inet6 hashsize 1024 maxelem 65536
Size in memory: 1240
References: 0
Number of entries: 0
Members:

Name: mwan3_dynamic_v4
Type: hash:net
Revision: 6
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 456
References: 0
Number of entries: 0
Members:

Name: mwan3_dynamic_v6
Type: hash:net
Revision: 6
Header: family inet6 hashsize 1024 maxelem 65536
Size in memory: 1240
References: 0
Number of entries: 0
Members:

Name: mwan3_custom_v4
Type: hash:net
Revision: 6
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 456
References: 0
Number of entries: 0
Members:

Name: mwan3_custom_v6
Type: hash:net
Revision: 6
Header: family inet6 hashsize 1024 maxelem 65536
Size in memory: 1240
References: 0
Number of entries: 0
Members:

Name: mwan3_connected
Type: list:set
Revision: 3
Header: size 8
Size in memory: 88
References: 0
Number of entries: 0
Members:
```
```shell
$ uci show opennds
opennds.@opennds[0]=opennds
opennds.@opennds[0].enabled='1'
opennds.@opennds[0].fwhook_enabled='1'
opennds.@opennds[0].allow_legacy_splash='1'
opennds.@opennds[0].gatewayinterface='br-lan'
opennds.@opennds[0].gatewayname='CA'
opennds.@opennds[0].gatewayfqdn='status.client'
opennds.@opennds[0].maxclients='250'
opennds.@opennds[0].preauthidletimeout='30'
opennds.@opennds[0].authidletimeout='120'
opennds.@opennds[0].sessiontimeout='1200'
opennds.@opennds[0].checkinterval='60'
opennds.@opennds[0].uploadrate='0'
opennds.@opennds[0].downloadrate='0'
opennds.@opennds[0].ratecheckwindow='2'
opennds.@opennds[0].uploadquota='0'
opennds.@opennds[0].downloadquota='0'
opennds.@opennds[0].authenticated_users='allow all'
opennds.@opennds[0].walledgarden_fqdn_list='baidu.com'
opennds.@opennds[0].walledgarden_port_list='443 80'
opennds.@opennds[0].users_to_router='allow tcp port 53' 'allow udp port 53' 'allow udp port 67' 'allow tcp port 22' 'allow tcp port 23' 'allow tcp port 80' 'allow tcp port 443'
```

**Comment by zedgu** [comment]
> @zedgu
> I have investigated a little. It seems Baidu is structured to use multiple FQDNs as well as direct IP address access.
> 
> You will have to list not only all the FQDNs it uses, but also manually add the ip addresses it uses without DNS.
> Usually the same FQDN and IP address information is used to block access, whereas you are trying to ALLOW access.
> 
> I suspect Baidu have done this intentionally to stop blocking - after all search engines generate a huge amount of revenue.
> **If I am correct, the problem that you have is that this will also stop you from allowing access in your walled garden.**

Baidu is just a sample here, any other domain is the same.

**Comment by bluewavenet** [comment]
@zedgu 
The Walled garden is in use in production at numerous sites that I know of, with no reported issues.
Common setups here are for Facebook and Paypal.

I did try baidu.com here on my test system. My findings so far are as follows:

1. baidu.com accesses the following FQDN's:
  * baidu.com
  * shifen.com
  * wshifen.com
  * gshifen.com
  * jiashule.com
  * jomodns.com
 
2. It also accesses numerous ip addresses that do not seem to reverse lookup - due to other commitments I have not had time to look into this further.

If you look at the system log just after restarting openNDS, do you see dnsmasq restarting? You should do...


None of your ipsets are populated - seems odd.
With openNDS running,  connect without logging in, then try to go to baidu.com in your browser.
Now run:
`ipset list walledgarden`

What do you get this time?

**Comment by zedgu** [comment]
system log:
```shell
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: openNDS Version 8.1.1
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Attempting to Bind to interface: br-lan
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Interface br-lan is up
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Interface br-lan is at 192.168.10.1 ()
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: ipset support is available
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: dnsmasq ipset support is available
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Created web server on 192.168.10.1:2050
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Legacy html Splash Page is Enabled.
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Initializing firewall rules
Thu May 13 17:21:38 2021 user.notice dnsmasq: DNS rebinding protection is active, will discard upstream RFC1918 responses!
Thu May 13 17:21:38 2021 user.notice dnsmasq: Allowing 127.0.0.0/8 responses
```

`ipset list walledgarden'
```shell
$ ipset list walledgarden
Name: walledgarden
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 256
References: 4
Number of entries: 1
Members:
xx.246.63.154
```

**Comment by zedgu** [comment]
system log:
```shell
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: openNDS Version 8.1.1
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Attempting to Bind to interface: br-lan
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Interface br-lan is up
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Interface br-lan is at 192.168.10.1 ()
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: ipset support is available
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: dnsmasq ipset support is available
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Created web server on 192.168.10.1:2050
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Legacy html Splash Page is Enabled.
Thu May 13 17:21:38 2021 daemon.notice opennds[10847]: Initializing firewall rules
Thu May 13 17:21:38 2021 user.notice dnsmasq: DNS rebinding protection is active, will discard upstream RFC1918 responses!
Thu May 13 17:21:38 2021 user.notice dnsmasq: Allowing 127.0.0.0/8 responses
```

`ipset list walledgarden'
```shell
$ ipset list walledgarden
Name: walledgarden
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 256
References: 4
Number of entries: 1
Members:
xx.246.63.154
```

**Comment by bluewavenet** [comment]
@zedgu 
I reran my test for comparison.

First I enabled dns logging in dnsmasq by adding the following to /etc/config/dhcp:
```
	option logdhcp '1'
	option logqueries '1'
	option logfacility '/tmp/dnsmasq.log'


```

Whilst logged in I went to `http://baidu.com`

Looking at the log I could see my laptop had called to resolve the following:
```
baidu.com
shifen.com
gshifen.com
wshifen.com
jaishule.com
jomodns.com
```
I then looked at the walled garden:

```
root@BlueWave-49510:~# ipset list
Name: walledgarden
Type: hash:ip
Revision: 4
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 2016
References: 1
Number of entries: 50
Members:
180.101.49.156
110.242.68.202
118.112.225.38
182.61.200.83
117.34.37.38
121.12.53.38
103.235.47.114
125.74.40.38
104.193.88.77
111.206.208.190
103.235.46.231
103.235.46.122
104.193.88.125
112.34.111.34
112.80.255.43
180.97.33.42
220.181.33.24
171.107.85.38
124.237.176.164
118.180.40.38
2.16.56.240
112.34.111.183
220.181.111.91
183.232.232.54
103.235.46.82
112.34.113.38
110.242.69.184
103.235.46.140
182.61.200.63
110.242.68.137
112.80.255.39
103.235.46.232
103.235.47.122
103.235.46.139
124.237.176.165
112.34.111.18
111.206.209.111
110.242.68.201
103.235.46.171
125.64.104.38
111.206.209.109
117.185.16.213
111.206.209.69
104.193.88.123
125.74.1.38
125.74.42.38
103.235.46.170
180.101.38.38
117.185.16.230
103.235.47.115
root@BlueWave-49510:~# 

```

I looked briefly at luci/status/realtime graphs/ and Connections tab.
I turned on DNS lookups.
I could see my laptop has also made TCP connections to ip addresses in similar ranges to those listed in the ipset, but these did not reverse lookup so would not be added into the ipset automatically. did not have time to make not of them all though.

I added the following to /etc/config/opennds:
`	list walledgarden_fqdn_list 'baidu.com shifen.com wshifen.com gshifen.com jiashule.com jomodns.com wwwbaidu.jomodns.com'`

Deauthenticating my laptop with ndsctl deauth [my ip address]
I then tried to browse to `http://baidu.com` and got a blank screen.

In summary,:
 * baidu.com uses multiple fqdns. This is normal, we see it here with Facebook, Paypal etc
 * baidu.com seems to have references to ip addresses directly and not by name. This is common where a site does not want ipset to be used for BLOCKING. Unfortunately is will make an ipset for ALLOWING by fqdn only useless.

**Comment by bluewavenet** [comment]
If anyone has a list of non-dns ips that baidu.com uses, please comment here.
